### ISSUE TITLE
As a user, I want to upload air quality and location data to the website

### ISSUE THREAD

**Issue by TangoYankee** [openoakland/woeip#81]
**Is your feature request related to a problem? Please describe.**
After collecting air quality data, I would like to upload it to WOAQ to be stored. 

**Describe the solution you'd like**
I will have two files:
1) A csv file which contains air quality measurements taken by a dustrak sensor
2) A log file which contains GPS location data taken by a ....gps?... @theecrit, what's the device?

As a signed in user, I can navigate to an `upload`/`add data` page. On this page, there is a button `find files on my computer`. Clicking this button will open a file explorer where I can select the correct files. I will receive feedback on whether the process was a success. If it was a failure, I will be told and given another chance.

**Describe alternatives you've considered**
Previous version had numerous fields on the form, each for a different value. However, having the user input each value is considered too tedious. It may even be overwhelming to ask so many questions. This method is preferred because the data will be parsed automatically.

**Additional context**
This issue is a prerequisite for issue #31. It provides the backed structure to the upload process.
In reference to the wireframes, it does not achieve the *drag and drop* functionality. However, it should achieve the *find files on my computer* functionality. Issue #63 depends on this process. It will be passed the URIs of the files in the S3 bucket.

**Labels**
- [x] Add a status label
  - `backlog`
  - `discussing`
  - `ready`
  - `implementing`
- [x] Add additional labels as needed

### For Developers
**Affected Components**
- `form` for upload
- `view` for upload
- `template` for upload

**Technical Resources**
[Django Forms](https://docs.djangoproject.com/en/2.1/topics/forms/)
[Django Messages Framework](https://docs.djangoproject.com/en/2.1/ref/contrib/messages/)
[Multiple File Upload tutorial](https://simpleisbetterthancomplex.com/tutorial/2016/11/22/django-multiple-file-upload-using-ajax.html)
[Multiple FIles Tutorial Source Code](https://github.com/sibtc/multiple-file-upload)
[Django Files Tutorial](https://simpleisbetterthancomplex.com/tutorial/2016/08/01/how-to-upload-files-with-django.html)

**Design Documentation**
[Wireframes V3](https://github.com/openoakland/woeip/files/2959738/WOEIP-UploadFlow-v3.pdf)

- Input for this issue: the two user-uploaded files
- Data processing in this issue: store the data files in the S3 bucket
- Output for this issue:
  - On success: generate uris to the locations of the files in the S3 bucket. Later, these uris will be passed to `data verification view` (issue #63) for processing 
  - On failure: generate as specific of an error message as possible, and return to the `add data` view. These error messages and their handling can be later refined as the javascript is added in issue #31.

**Comment by theecrit** [comment]
@TangoYankee Under "Additional Context," you mention that "This issue is a prerequisite for issue #31. It provides the backed structure to the upload process." Can you clarify the relationship between this issue and https://github.com/openoakland/woeip/issues/43? My gut says that #43 needs to be clarified before it's actionable, and so that it's mutually exclusive from this Issue.

**Comment by TangoYankee** [comment]
100% #43, dustrak upload needs to be re-written and clarified. It's a remnant from before we had the proper workflow. 

This issue (#81) is a pre-requisite for #43. Under #43, we will extract the meta data of the file and send it to the verification view (#63). We might actually want to consider combining #43 and #63.

**Comment by Ethan-bradley** [comment]
@nbilenko Currently ajax is being used to send the data. Does that make sense for what data management you want to do?

**Comment by theecrit** [comment]
Just want to check in on the progress of this issue (and its related https://github.com/openoakland/woeip/issues/31). @Ethan-bradley @nbilenko What are next steps for this one? Thanks!

**Comment by theecrit** [comment]
@TangoYankee Just assigned this to you; you're up.

**Comment by nbilenko** [comment]
@TangoYankee Iterated on your already substantial progress, result is here: https://github.com/openoakland/woeip/tree/nb/dragdrop6
I think we're close to done on this issue.

Screenshots attached. 

Users initiate the upload by selecting files (single or multiple) using the button or by dragging anywhere on page, and then pressing the "Upload" button to confirm. Only .csv and .log files can be uploaded, up to 100Kb.

All uploaded and not yet processed files by this user appear in the list below, and the users can delete uploaded files.

Once the user decides on the files, they press the Submit session button, which should redirect to the verify view (that processes the file data and marks the files as processed, creates a new session, and presents metadata to the user for checking, as described in https://github.com/openoakland/woeip/issues/90). Currently the mockup view returns a "success" http response.

<img width="835" alt="Screen Shot 2019-06-03 at 3 17 28 PM" src="https://user-images.githubusercontent.com/4154263/58838138-b8d4d500-8612-11e9-95f3-2c66c0d385ca.png">

<img width="782" alt="Screen Shot 2019-06-03 at 3 19 00 PM" src="https://user-images.githubusercontent.com/4154263/58838269-0bae8c80-8613-11e9-94b7-7bb326ef5d24.png">

**Comment by theecrit** [comment]
Wow, looking really good! If I understand correctly, this now requires the user to 1) verify the files after selecting them to initiate the upload, *and* 2) submit the files once they've been uploaded.

This seems like a lot of steps for the user.

The original intention was for the files to begin uploading as soon as the user does one of two things:
1. Releasing dragged files over the drop zone, or...
2. Clicking the "Select" button once files are selected, if using the file browser option. 

Is this possible? It seems like overkill to make them verify the files *prior* to upload, submit files *after upload but before processing*, and then finally *verify the data after upload and processing*.

**Comment by TangoYankee** [comment]
This provides feedback that the correct files were selected. It also tells the user that their first file didn't disappear down a black hole.

Perhaps most importantly, it prevents a cascade of errors that we would need to spend a couple more weeks having to design catches for.

That said, the screen shots look like the user presses `upload` to get it added to the list. We could look into cutting that part out. The essential bit is keeping the `submit session` bit.

**Comment by theecrit** [comment]
This implementation is a very different experience than what had been discussed based on the wireframes. If there are concerns about the design, we should discuss that rather than re-engineer the experience without discussion.

The intent of the experience is for the user to be able to select two files *at the same time*, and initiate upload by either: 1) dragging *both files together* and releasing over the drop zone, or 2) clicking the button that reveals the file browser and selecting the file pair there, then clicking the "select" button. It's very possible that I'm forgetting a conversation, but when did this change?

**Comment by TangoYankee** [comment]
1) This is the conversation where it changed. This is just a dev branch. It's not even a pull on master.

2) Not every detail can be talked about and documented. Otherwise, nothing will ever get done. We'll be talking until 2021.

3) We're volunteers with day jobs. Compromises must be made to keep a reasonable technical scope.

**Comment by nbilenko** [comment]
Agreed that it would be good to discuss what the flow should be. Having the file review instead of an immediate upload is a less straightforward implementation technically, but Tim found a great resource that had that option, so that's what he and I adapted. 
There's an option to turn the review off and have the files upload immediately with this implementation as well. I thought it might be good to try the review flow first in case it may not have been known to everyone that it's doable without a lot of extra work (it wasn't to me). Sorry for not making it clear that that's why it's different from the wireframe - I was hoping it would be a discussion starter. 
In my experience when testing it, it seemed useful to have the option to upload one file at a time (e.g. if the files are in different places on the computer) and then submit. But I recognize the upside of having a more streamlined flow as well.

**Comment by theecrit** [comment]
Thanks for the context, @nbilenko. Let's just be sure to document these kinds of decisions on the Issues, so folks understand why decisions are being made.

**Comment by TangoYankee** [comment]
Minor bookkeeping: this issue is for the backend collection of the files. The Drag and Drop UI functionality is issue #31. No biggie ¯\_(ツ)_/¯  

Anyway, I haven't been able to get the Drag and Drop functionality working. It seems it will take more research. I pivoted to the backend collection of the csv and log files. The user interface will need to be addressed before launch. However, this solution will allow downstream features to be built.

The branch is [ty/upload-sep-fields](https://github.com/openoakland/woeip/tree/ty/upload-sep-fields). ~I'll write a full debrief of decisions this evening. Until then, feel free to look it over and cast judgement~ 
Debrief is in PR #109

**Comment by theecrit** [comment]
Closing as "won't fix" due to tech stack updates.
